rust/protocol/src/crypto.rs

```diff
@@ -1,156 +1,244 @@
-    mac_key: &[u8],
-) -> Result<Vec<u8>, EncryptionError> {
-    let mut ctext = aes_256_ctr_encrypt(msg, cipher_key)?;
-    let mac = hmac_sha256(mac_key, &ctext);
-    ctext.extend_from_slice(&mac[..10]);
-    Ok(ctext)
-}
+    /// Encrypt plaintext `msg` with AES-256 and embed a computed HMAC into the returned bytes.
+    ///
+    /// *Implementation note: within the body of this method, only the first [`MAC_KEY_LENGTH`]
+    /// bytes of the computed MAC are used.*
+    pub fn aes_256_ctr_hmac_encrypt(
+        msg: &[u8],
+        cipher_key: &[u8; AES_256_KEY_SIZE],
+        mac_key: &[u8; MAC_KEY_LENGTH],
+    ) -> Vec<u8> {
+        let mut ctext = aes_256_ctr_encrypt(msg, cipher_key);
+        let mac = hmac_sha256(mac_key, &ctext);
+        ctext.extend_from_slice(&mac[..MAC_KEY_LENGTH]);
+        ctext
+    }
 
-pub(crate) fn aes256_ctr_hmacsha256_decrypt(
-    ctext: &[u8],
-    cipher_key: &[u8],
-    mac_key: &[u8],
-) -> Result<Vec<u8>, DecryptionError> {
-    if ctext.len() < 10 {
-        return Err(DecryptionError::BadCiphertext("truncated ciphertext"));
+    /// Validate the HMAC `mac_key` against the ciphertext `ctext`, then decrypt `ctext` using
+    /// AES-256 with `cipher_key` and [`aes_256_ctr_decrypt`].
+    ///
+    /// *Implementation note: the last [`MAC_KEY_LENGTH`] bytes of the `ctext` slice represent the
+    /// truncated HMAC of the rest of the message, as generated by [`aes_256_ctr_hmac_encrypt`].*
+    pub fn aes_256_ctr_hmac_decrypt(
+        ctext: &[u8],
+        cipher_key: &[u8; AES_256_KEY_SIZE],
+        mac_key: &[u8; MAC_KEY_LENGTH],
+    ) -> Result<Vec<u8>, DecryptionError> {
+        if ctext.len() < MAC_KEY_LENGTH {
+            return Err(DecryptionError("truncated ciphertext"));
+        }
+        let (ctext, ctext_mac) = ctext.split_at(ctext.len() - MAC_KEY_LENGTH);
+        let ctext_mac = array_ref![ctext_mac, 0, MAC_KEY_LENGTH];
+        let our_mac = hmac_sha256(mac_key, ctext);
+        let our_mac = array_ref![&our_mac, 0, MAC_KEY_LENGTH];
+        let same: bool = our_mac.ct_eq(ctext_mac).into();
+        if !same {
+            return Err(DecryptionError("MAC verification failed"));
+        }
+        Ok(aes_256_ctr_decrypt(ctext, cipher_key))
     }
-    let ptext_len = ctext.len() - 10;
-    let our_mac = hmac_sha256(mac_key, &ctext[..ptext_len]);
-    let same: bool = our_mac[..10].ct_eq(&ctext[ptext_len..]).into();
-    if !same {
-        return Err(DecryptionError::BadCiphertext("MAC verification failed"));
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        use std::convert::TryInto;
+
+        use rand::rngs::OsRng;
+        use rand::Rng;
+
+        #[test]
+        fn static_test() {
+            let key: [u8; AES_256_KEY_SIZE] =
+                hex::decode("603DEB1015CA71BE2B73AEF0857D77811F352C073B6108D72D9810A30914DFF4")
+                    .expect("valid hex")
+                    .try_into()
+                    .expect("correct array size");
```

Comment:
The `hex_literal` crate used elsewhere in the workspace can check these conditions at compile time; if you're adding more tests it might be a good time to switch over (incrementally).